### PR TITLE
feat(ui): Display pretty cron schedule

### PR DESCRIPTION
**Pull request by alexec**
image

image

image

image

### CODE REVIEW

**Comment by simster7** [comment]
Ah, I though you had meant a cron scheduler builder. Not one to simply display a humanized version :)

**Comment by alexec** [comment]
> Ah, I though you had meant a cron scheduler builder. Not one to simply display a humanized version :)

A builder could be a v2. We've both looked for libraries, and not found a good builder library.

**Comment by alexec** [comment]
@simster7 could I please request your review?

File: ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx [hunk 1]
```diff
@@ -137,8 +138,10 @@ export class CronWorkflowList extends BasePage<RouteComponentProps<any>, State>
                             <div className='columns small-1'>{w.spec.suspend ? <i className='fa fa-pause' /> : <i className='fa fa-clock' />}</div>
                             <div className='columns small-3'>{w.metadata.name}</div>
                             <div className='columns small-2'>{w.metadata.namespace}</div>
-                            <div className='columns small-2'>{w.spec.schedule}</div>
-                            <div className='columns small-2'>
+                            <div className='columns small-3'>
+                                <Schedule schedule={w.spec.schedule} />
```
Comment (simster7) [inline ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:142]: I don't think the "pretty" schedule should be displayed in the list view. It's not necessary when looking at all cron workflows (as opposed to just one in detail) and from your screenshot it looks like they're frequently cut off (bad UX)

File: ui/src/app/cron-workflows/components/cron-workflow-spec-editior.tsx [hunk 2]
```diff
@@ -12,6 +13,7 @@ export const CronWorkflowSpecEditor = ({onChange, spec}: {spec: CronWorkflowSpec
                     <div className='columns small-3'>Schedule</div>
                     <div className='columns small-9'>
                         <TextInput value={spec.schedule} onChange={schedule => onChange({...spec, schedule})} />
+                        <ScheduleValidator schedule={spec.schedule} />
```
Comment (simster7) [inline ui/src/app/cron-workflows/components/cron-workflow-spec-editior.tsx:16]: I think this should be inline with the input, not in a new line. The spec string will never be long enough to occupy even a tenth of the length of the input box, so no reason we can't use the remaining length more effectively with the validator

**Review by simster7 — CHANGES_REQUESTED** [review]

File: ui/src/app/cron-workflows/components/schedule.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    I think we must assume that these libraries (or any two libraries) will never be exactly the same and accept that
+    sometime it'll not work as expected. Therefore, we must let the user know about this.
+ */
+
+export const Schedule = ({schedule}: {schedule: string}) => {
+    try {
+        const pretty = x.toString(schedule);
+        return (
+            <span title={pretty}>
+                {schedule} <small>({pretty})</small>
```
Comment (simster7) [inline ui/src/app/cron-workflows/components/schedule.tsx:20]: Not sure if parenthesis is my favorite way to segment them. Perhaps
```suggestion
                <code>{schedule}</code> {pretty}
```
?

File: ui/src/app/cron-workflows/components/cron-workflow-spec-editior.tsx (hunk 2 above)
Comment (alexec) [reply ui/src/app/cron-workflows/components/cron-workflow-spec-editior.tsx:16]: I'd like to leave this, it is a small panel and I don't think it should be optimised.

File: ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx (hunk 1 above)
Comment (alexec) [reply ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:142]: I'd like this to be visible in the list. My reasons:

* Commonly I believe that people use simple schedules that are short.
* It is easier for users to visually parse, vs raw schedule.
* I expect truncation to happen about. 

More complex example:

image

I'm going to ask the community.

Comment (alexec) [reply ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:142]: I can see a bug in the library. `0 12 * */1 *`. This should be "At 12:00 in every month.”

Comment (alexec) [reply ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:142]: huh - 1st of the month?

Comment (alexec) [reply ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:142]: I *think* "*/1" === "*"?

**Review by simster7 — APPROVED** [review]
Pre approval

**Comment by alexec** [comment]
I've pork-barrelled changes to formatting durations. We can rollback if disliked

**Pull request merged**